nullaway/src/main/java/com/uber/nullaway/NullMarkedCache.java

```diff
@@ -0,0 +1,96 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway;
+
+import static com.uber.nullaway.NullabilityUtil.NULLMARKED;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.tools.javac.code.Symbol;
+
+/**
+ * A helper cache class to keep track of (inner) classes marked with @NullMarked and to resolve
+ * their outermost class. Used by {@link NullabilityUtil}.
+ */
+final class NullMarkedCache {
+  private static final int MAX_CACHE_SIZE = 200;
+
+  private final Cache<Symbol.ClassSymbol, NullMarkedCacheRecord> innerCache =
+      CacheBuilder.newBuilder().maximumSize(MAX_CACHE_SIZE).build();
+
+  public NullMarkedCache() {}
+
+  /**
+   * Retrieve the (outermostClass, isNullMarked) record for a given class symbol.
+   *
+   * <p>This method is recursive, using the cache on the way up and populating it on the way down.
+   *
+   * @param classSymbol The class to query, possibly an inner class
+   * @return A record including the outermost class in which the given class is nested, as well as
+   *     boolean flag noting whether itself or any of the classes containing it are @NullMarked (See
+   *     Record)
+   */
+  public NullMarkedCacheRecord get(Symbol.ClassSymbol classSymbol) {
+    NullMarkedCacheRecord record = innerCache.getIfPresent(classSymbol);
+    if (record != null) {
+      return record;
+    }
+    if (classSymbol.getNestingKind().isNested()) {
+      Symbol.ClassSymbol enclosingClass = ASTHelpers.enclosingClass(classSymbol);
+      // enclosingSymbol can be null in weird cases like for array methods
+      if (enclosingClass != null) {
+        NullMarkedCacheRecord recordForEnclosing = get(enclosingClass);
+        record =
+            new NullMarkedCacheRecord(
+                recordForEnclosing.outermostClassSymbol,
+                recordForEnclosing.isNullMarked || classSymbol.getAnnotation(NULLMARKED) != null);
+      }
+    }
+    if (record == null) {
+      // We are already at the outermost class (we can find), so let's create a record for it
+      record =
+          new NullMarkedCacheRecord(classSymbol, classSymbol.getAnnotation(NULLMARKED) != null);
+    }
+    innerCache.put(classSymbol, record);
+    return record;
+  }
+
+  /**
+   * Immutable record holding the outermost class symbol and @NullMarked state (annotated or not)
```

Comment:
@lazaroclapp do we still have a need to cache the outermost class symbol for the purposes of `@NullMarked` support?  It seems like that cache is being used for things unrelated to `@NullMarked`.  If we don't need to cache the symbol, it would simplify making the `NullMarkedCache` non-static

Reply:
So, this is used by `fromAnnotatedPackage(...)` both to check if the package of `outermostClassSymbol` is explicitly passed to NullAway as an annotated package as a command line flag, as well as to check for `@NullMarked` at the package level. 

I think matchers like `fromExplicitlyAnnotatedPackage(...)` could always be made to operate on any `classSymbol`, though, not just the outermost one. I am not sure if we need to have the outermost class symbol in order for `symbol.packge().getAnnotation(NULLMARKED)` to work. But, even so, we could always go back to manually calling:

```
Symbol.ClassSymbol outermostClassSymbol = getOutermostClassSymbol(symbol);
```

which is what we had anyways before this PR.

Reply:
I see.  The thing I forgot is that checking for presence in an "annotated package" is now more involved since we also need to check for a `package-info.java` with the annotation, which requires traversing the symbol structure.  Still not 100% sure we need to cache the outermost class symbol though; will think more.

Reply:
Note that we need to traverse the symbol structure either way, because we can be looking at third-party code, or just code outside of the current compilation, where some inner classes are marked `@NullMarked` and others aren't (in fact, we need to do this for `MethodSymbol`s as well)

Reply:
FWIW, I was not planning on extending this PR to handle `@NullMarked` on methods.  I don't think support for that is finalized in JSpecify, so I think it can wait and be handled in a follow-up PR.